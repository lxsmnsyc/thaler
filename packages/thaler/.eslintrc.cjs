module.exports = {
  "root": true,
  "extends": [
    'lxsmnsyc/typescript',
  ],
  "parserOptions": {
    "project": "./tsconfig.eslint.json",
    "tsconfigRootDir": __dirname,
  },
  "rules": {
    "import/no-extraneous-dependencies": [
      "error", {
        "devDependencies": ["**/*.test.ts"]
      }
    ]
    "@typescript-eslint/no-unsafe-argument": "off",
    "@typescript-eslint/no-unsafe-assignment": "off",
    "@typescript-eslint/no-unsafe-return": "off"
  }
};