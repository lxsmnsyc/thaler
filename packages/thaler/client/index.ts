import { createReference, deserialize, toJSONAsync } from 'seroval';
import ThalerError from '../shared/error';
import type {
  ThalerPostInit,
  ThalerPostParam,
  ThalerFunctionInit,
  ThalerFunctions,
  ThalerFunctionTypes,
  ThalerGetInit,
  ThalerGetParam,
  MaybePromise,
} from '../shared/types';
import {
  XThalerID,
  XThalerInstance,
  patchHeaders,
  serializeFunctionBody,
  toFormData,
  toURLSearchParams,
} from '../shared/utils';

interface HandlerRegistrationResult {
  type: ThalerFunctionTypes;
  id: string;
}

export function $$server(id: string): HandlerRegistrationResult {
  return { type: 'server', id };
}
export function $$post(id: string): HandlerRegistrationResult {
  return { type: 'post', id };
}
export function $$get(id: string): HandlerRegistrationResult {
  return { type: 'get', id };
}
export function $$fn(id: string): HandlerRegistrationResult {
  return { type: 'fn', id };
}
export function $$pure(id: string): HandlerRegistrationResult {
  return { type: 'pure', id };
}
export function $$loader(id: string): HandlerRegistrationResult {
  return { type: 'loader', id };
}
export function $$action(id: string): HandlerRegistrationResult {
  return { type: 'action', id };
}

export type Interceptor = (request: Request) => MaybePromise<Request>;

const INTERCEPTORS: Interceptor[] = [];

export function interceptRequest(callback: Interceptor): void {
  INTERCEPTORS.push(callback);
}

async function serverHandler(
  type: ThalerFunctionTypes,
  id: string,
  init: RequestInit,
): Promise<Response> {
  patchHeaders(type, id, init);
  let root = new Request(id, init);
  for (const intercept of INTERCEPTORS) {
    root = await intercept(root);
  }
  const result = await fetch(root);
  return result;
}

async function postHandler<P extends ThalerPostParam>(
  id: string,
  form: P,
  init: ThalerPostInit = {},
): Promise<Response> {
  return await serverHandler('post', id, {
    ...init,
    method: 'POST',
    body: toFormData(form),
  });
}

async function getHandler<P extends ThalerGetParam>(
  id: string,
  search: P,
  init: ThalerGetInit = {},
): Promise<Response> {
  return await serverHandler(
    'get',
    `${id}?${toURLSearchParams(search).toString()}`,
    {
      ...init,
      method: 'GET',
    },
  );
}

declare const $R: Record<string, unknown>;

class SerovalChunkReader {
  private reader: ReadableStreamDefaultReader<Uint8Array>;
  private buffer = '';
  private done = false;

  constructor(stream: ReadableStream<Uint8Array>) {
    this.reader = stream.getReader();
  }

  async readChunk(): Promise<void> {
    // if there's no chunk, read again
    const chunk = await this.reader.read();
    if (chunk.done) {
      this.done = true;
    } else {
      // repopulate the buffer
      this.buffer += new TextDecoder().decode(chunk.value);
    }
  }

  async next(): Promise<IteratorResult<string>> {
    // Check if the buffer is empty
    if (this.buffer === '') {
      // if we are already done...
      if (this.done) {
        return {
          done: true,
          value: undefined,
        };
      }
      // Otherwise, read a new chunk
      await this.readChunk();
      return await this.next();
    }
    // Read the "byte header"
    // The byte header tells us how big the expected data is
    // so we know how much data we should wait before we
    // deserialize the data
    const bytes = Number.parseInt(this.buffer.substring(1, 11), 16); // ;0x00000000;
    // Check if the buffer has enough bytes to be parsed
    while (bytes > this.buffer.length - 12) {
      // If it's not enough, and the reader is done
      // then the chunk is invalid.
      if (this.done) {
        throw new Error('Malformed server function stream.');
      }
      // Otherwise, we read more chunks
      await this.readChunk();
    }
    // Extract the exact chunk as defined by the byte header
    const partial = this.buffer.substring(12, 12 + bytes);
    // The rest goes to the buffer
    this.buffer = this.buffer.substring(12 + bytes);
    // Deserialize the chunk
    return {
      done: false,
      value: deserialize(partial),
    };
  }

  async drain(): Promise<void> {
    while (true) {
      const result = await this.next();
      if (result.done) {
        break;
      }
    }
  }
}

async function deserializeStream<T>(
  id: string,
  response: Response,
): Promise<T> {
  const instance = response.headers.get(XThalerInstance);
  const target = response.headers.get(XThalerID);
  if (!instance || target !== id) {
    throw new Error(`Invalid response for ${id}.`);
  }
  if (!response.body) {
    throw new Error('missing body');
  }
  const reader = new SerovalChunkReader(response.body);

  const result = await reader.next();

  if (!result.done) {
    reader.drain().then(
      () => {
        delete $R[instance];
      },
      () => {
        // no-op
      },
    );
  }

  if (response.ok) {
    return result.value as T;
  }
  if (import.meta.env.DEV) {
    throw result.value;
  }
  throw new ThalerError(id);
}

async function fnHandler<T, R>(
  id: string,
  scope: () => unknown[],
  value: T,
  init: ThalerFunctionInit = {},
): Promise<R> {
  return deserializeStream(
    id,
    await serverHandler('fn', id, {
      ...init,
      method: 'POST',
      body: await serializeFunctionBody({
        scope: scope(),
        value,
      }),
    }),
  );
}

async function pureHandler<T, R>(
  id: string,
  value: T,
  init: ThalerFunctionInit = {},
): Promise<R> {
  return deserializeStream(
    id,
    await serverHandler('pure', id, {
      ...init,
      method: 'POST',
      body: JSON.stringify(await toJSONAsync(value)),
    }),
  );
}

async function loaderHandler<P extends ThalerGetParam, R>(
  id: string,
  search: P,
  init: ThalerGetInit = {},
): Promise<R> {
  return deserializeStream<R>(
    id,
    await serverHandler(
      'loader',
      `${id}?${toURLSearchParams(search).toString()}`,
      {
        ...init,
        method: 'GET',
      },
    ),
  );
}

async function actionHandler<P extends ThalerPostParam, R>(
  id: string,
  form: P,
  init: ThalerPostInit = {},
): Promise<R> {
  return deserializeStream<R>(
    id,
    await serverHandler('action', id, {
      ...init,
      method: 'POST',
      body: toFormData(form),
    }),
  );
}

export function $$clone(
  { type, id }: HandlerRegistrationResult,
  scope: () => unknown[],
): ThalerFunctions {
  switch (type) {
    case 'server':
      return Object.assign(serverHandler.bind(null, 'server', id), {
        type,
        id,
      });
    case 'post':
      return Object.assign(postHandler.bind(null, id), {
        type,
        id,
      });
    case 'get':
      return Object.assign(getHandler.bind(null, id), {
        type,
        id,
      });
    case 'fn':
      return Object.assign(fnHandler.bind(null, id, scope), {
        type,
        id,
      });
    case 'pure':
      return Object.assign(pureHandler.bind(null, id), {
        type,
        id,
      });
    case 'loader':
      return Object.assign(loaderHandler.bind(null, id), {
        type,
        id,
      });
    case 'action':
      return Object.assign(actionHandler.bind(null, id), {
        type,
        id,
      });
    default:
      throw new Error('unknown registration type');
  }
}

export function $$ref<T>(id: string, value: T): T {
  return createReference(`thaler--${id}`, value);
}
