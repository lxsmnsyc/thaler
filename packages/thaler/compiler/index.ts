import * as babel from '@babel/core';
import path from 'path';
import type { PluginOptions } from './plugin';
import thalerBabel from './plugin';

export type Options = Omit<PluginOptions, 'source'>;

export interface CompileResult {
  code: string;
  map: babel.BabelFileResult['map'];
}

export async function compile(
  id: string,
  code: string,
  options: Options,
): Promise<CompileResult> {
  const pluginOption = [
    thalerBabel,
    {
      source: id,
      prefix: options.prefix,
      mode: options.mode,
      functions: options.functions,
    },
  ];
  const plugins: NonNullable<
    NonNullable<babel.TransformOptions['parserOpts']>['plugins']
  > = ['jsx'];
  if (/\.[mc]?tsx?$/i.test(id)) {
    plugins.push('typescript');
  }
  const result = await babel.transformAsync(code, {
    plugins: [pluginOption],
    parserOpts: {
      plugins,
    },
    filename: path.basename(id),
    ast: false,
    sourceMaps: true,
    configFile: false,
    babelrc: false,
    sourceFileName: id,
  });

  if (result) {
    return {
      code: result.code || '',
      map: result.map,
    };
  }
  throw new Error('invariant');
}
