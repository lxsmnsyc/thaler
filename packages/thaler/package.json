{
  "name": "thaler",
  "version": "0.8.1",
  "type": "module",
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=10"
  },
  "license": "MIT",
  "keywords": [
    "pridepack"
  ],
  "devDependencies": {
    "@types/babel__core": "^7.20.3",
    "@types/babel__traverse": "^7.20.3",
    "@types/node": "^20.8.7",
    "eslint": "^8.54.0",
    "eslint-config-lxsmnsyc": "^0.6.5",
    "pridepack": "2.5.1",
    "tslib": "^2.6.2",
    "typescript": "^5.3.2",
    "vitest": "^0.34.6"
  },
  "dependencies": {
    "@babel/core": "^7.23.2",
    "@babel/helper-module-imports": "^7.22.15",
    "@babel/traverse": "^7.23.2",
    "@babel/types": "^7.23.0",
    "seroval": "^0.13.2"
  },
  "scripts": {
    "prepublishOnly": "pridepack clean && pridepack build",
    "build": "pridepack build",
    "type-check": "pridepack check",
    "lint": "pridepack lint",
    "clean": "pridepack clean",
    "watch": "pridepack watch",
    "start": "pridepack start",
    "dev": "pridepack dev",
    "test": "vitest"
  },
  "description": "Isomorphic server-side functions",
  "repository": {
    "url": "https://github.com/lxsmnsyc/thaler.git",
    "type": "git"
  },
  "homepage": "https://github.com/lxsmnsyc/thaler/tree/main/packages/thaler",
  "bugs": {
    "url": "https://github.com/lxsmnsyc/thaler/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Alexis Munsayac",
  "private": false,
  "typesVersions": {
    "*": {
      "compiler": [
        "./dist/types/compiler/index.d.ts"
      ],
      "client": [
        "./dist/types/client/index.d.ts"
      ],
      "server": [
        "./dist/types/server/index.d.ts"
      ],
      "utils": [
        "./dist/types/utils/index.d.ts"
      ]
    }
  },
  "types": "./dist/types/src/index.d.ts",
  "main": "./dist/cjs/production/index.cjs",
  "module": "./dist/esm/production/index.mjs",
  "exports": {
    ".": {
      "development": {
        "require": "./dist/cjs/development/index.cjs",
        "import": "./dist/esm/development/index.mjs"
      },
      "require": "./dist/cjs/production/index.cjs",
      "import": "./dist/esm/production/index.mjs",
      "types": "./dist/types/src/index.d.ts"
    },
    "./compiler": {
      "development": {
        "require": "./dist/cjs/development/compiler.cjs",
        "import": "./dist/esm/development/compiler.mjs"
      },
      "require": "./dist/cjs/production/compiler.cjs",
      "import": "./dist/esm/production/compiler.mjs",
      "types": "./dist/types/compiler/index.d.ts"
    },
    "./client": {
      "development": {
        "require": "./dist/cjs/development/client.cjs",
        "import": "./dist/esm/development/client.mjs"
      },
      "require": "./dist/cjs/production/client.cjs",
      "import": "./dist/esm/production/client.mjs",
      "types": "./dist/types/client/index.d.ts"
    },
    "./server": {
      "development": {
        "require": "./dist/cjs/development/server.cjs",
        "import": "./dist/esm/development/server.mjs"
      },
      "require": "./dist/cjs/production/server.cjs",
      "import": "./dist/esm/production/server.mjs",
      "types": "./dist/types/server/index.d.ts"
    },
    "./utils": {
      "development": {
        "require": "./dist/cjs/development/utils.cjs",
        "import": "./dist/esm/development/utils.mjs"
      },
      "require": "./dist/cjs/production/utils.cjs",
      "import": "./dist/esm/production/utils.mjs",
      "types": "./dist/types/utils/index.d.ts"
    }
  }
}
