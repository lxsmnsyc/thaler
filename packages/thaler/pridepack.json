{
  "target": "es2018",
  "entrypoints": {
    ".": "src/index.ts",
    "./compiler": "compiler/index.ts",
    "./client": "client/index.ts",
    "./server": "server/index.ts",
    "./utils": "utils/index.ts"
  }
}
