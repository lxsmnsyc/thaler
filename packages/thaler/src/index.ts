import type {
  ThalerPostFunction,
  ThalerPostHandler,
  ThalerPostParam,
  ThalerFunction,
  ThalerFnHandler,
  ThalerGetFunction,
  ThalerGetHandler,
  ThalerGetParam,
  ThalerPureFunction,
  ThalerPureHandler,
  ThalerServerFunction,
  ThalerServerHandler,
  ThalerLoaderHandler,
  ThalerLoaderFunction,
  ThalerActionHandler,
  ThalerActionFunction,
} from '../shared/types';

export * from '../shared/types';

export function server$(_handler: ThalerServerHandler): ThalerServerFunction {
  throw new Error('server$ cannot be called during runtime.');
}

export function post$<P extends ThalerPostParam>(
  _handler: ThalerPostHandler<P>,
): ThalerPostFunction<P> {
  throw new Error('post$ cannot be called during runtime.');
}

export function get$<P extends ThalerGetParam>(
  _handler: ThalerGetHandler<P>,
): ThalerGetFunction<P> {
  throw new Error('get$ cannot be called during runtime.');
}

export function fn$<T, R>(
  _handler: ThalerFnHandler<T, R>,
): ThalerFunction<T, R> {
  throw new Error('fn$ cannot be called during runtime.');
}

export function pure$<T, R>(
  _handler: ThalerPureHandler<T, R>,
): ThalerPureFunction<T, R> {
  throw new Error('pure$ cannot be called during runtime.');
}

export function loader$<P extends ThalerGetParam, R>(
  _handler: ThalerLoaderHandler<P, R>,
): ThalerLoaderFunction<P, R> {
  throw new Error('fn$ cannot be called during runtime.');
}

export function action$<P extends ThalerPostParam, R>(
  _handler: ThalerActionHandler<P, R>,
): ThalerActionFunction<P, R> {
  throw new Error('pure$ cannot be called during runtime.');
}

export function ref$<T>(_value: T): T {
  throw new Error('ref$ cannot be called during runtime.');
}

export {
  fromFormData,
  fromURLSearchParams,
} from '../shared/utils';
