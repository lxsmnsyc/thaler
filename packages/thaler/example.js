import compile from 'thaler/compiler';

const serverOptions = {
  prefix: 'example',
  mode: 'server',
};

const FILE = 'src/index.ts';

const code = `
import { createResource, createSignal, Suspense } from 'solid-js';
import { fn$, ref$ } from 'thaler';

const sleep = (ms: number) => new Promise((res) => {
  setTimeout(res, ms, true);
});

const sleepingValue = ref$(async (value: number) => {
  await sleep(1000);
  console.log('Received', value);
  return value;
});

export default function Example() {
  const [state, setState] = createSignal(0);

  const prefix = 'Server Count';

  const serverCount = fn$(async ([cb, value]: [typeof sleepingValue, number]) => (
    \`\${prefix}: \${await cb(value)}\`
  ));

  const [data] = createResource(state, (value) => serverCount([sleepingValue, value]));

  function increment() {
    setState((c) => c + 1);
  }

  return (
    <>
      <button onClick={increment}>
        {\`Client Count: \${state()}\`}
      </button>
      <div>
        <Suspense fallback={<h1>Loading</h1>}>
          <h1>{data()}</h1>
        </Suspense>
      </div>
    </>
  );
}

`;

console.log((await compile(FILE, code, serverOptions)).code);
