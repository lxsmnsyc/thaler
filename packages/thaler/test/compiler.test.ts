import { describe, expect, it } from 'vitest';
import type { Options } from '../compiler';
import compile from '../compiler';

const functions: Options['functions'] = [
  {
    name: 'example$',
    scoping: true,
    target: {
      name: 'example$',
      source: 'example-server-function',
      kind: 'named',
    },
    server: {
      name: '$$example',
      source: 'example-server-function/server',
      kind: 'named',
    },
    client: {
      name: '$$example',
      source: 'example-server-function/client',
      kind: 'named',
    },
  },
];

const serverOptions: Options = {
  prefix: 'example',
  mode: 'server',
  functions,
};

const clientOptions: Options = {
  prefix: 'example',
  mode: 'client',
  functions,
};

const FILE = 'src/index.ts';

describe('server$', () => {
  it('should transform', async () => {
    const code = `
import { server$ } from 'thaler';

const example = server$((request) => {
  return new Response('Hello World', {
    headers: {
      'content-type': 'text/html',
    },
    status: 200,
  });
});
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});
describe('get$', () => {
  it('should transform', async () => {
    const code = `
import { get$ } from 'thaler';

const example = get$(({ greeting, receiver}) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html',
    },
    status: 200,
  });
});
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});
describe('post$', () => {
  it('should transform', async () => {
    const code = `
import { post$ } from 'thaler';

const example = post$(({ greeting, receiver }) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html',
    },
    status: 200,
  });
});
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});
describe('fn$', () => {
  it('should transform', async () => {
    const code = `
import { fn$ } from 'thaler';

const PREFIX = 'Message: ';

const example = fn$(({ greeting, receiver }) => {
  const message = PREFIX + greeting + ', ' + receiver + '!';
  return message;
});
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
  it('should transform with local scope', async () => {
    const code = `
import { fn$ } from 'thaler';

function test() {
  const PREFIX = 'Message: ';
  
  const example = fn$(({ greeting, receiver }) => {
    const message = PREFIX + greeting + ', ' + receiver + '!';
    return message;
  });
}
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});

describe('pure$', () => {
  it('should transform', async () => {
    const code = `
import { pure$ } from 'thaler';

const sleep = (ms) => new Promise((res) => {
  setTimeout(res, ms, true);
});

const example = pure$(async ({ greeting, receiver }) => {
  await sleep(1000);
  const message = greeting + ', ' + receiver + '!';
  return message;
});
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});

describe('ref$', () => {
  it('should transform', async () => {
    const code = `
import { ref$ } from 'thaler';

const example = ref$(() => 'Hello World');
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});

describe('loader$', () => {
  it('should transform', async () => {
    const code = `
import { loader$ } from 'thaler';

const example = loader$(async ({ greeting, receiver }) => {
  const message = greeting + ', ' + receiver + '!';
  return message;
});
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});

describe('action$', () => {
  it('should transform', async () => {
    const code = `
import { action$ } from 'thaler';

const example = action$(async ({ greeting, receiver }) => {
  const message = greeting + ', ' + receiver + '!';
  return message;
});
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});

describe('custom server function', () => {
  it('should transform', async () => {
    const code = `
import { example$ } from 'example-server-function';

function exampleProgram() {
  const greeting = 'Hello';
  const receiver = 'World';
  
  const example = example$(() => {
    const message = greeting + ', ' + receiver + '!';
    return message;
  });
}
  `;
    expect((await compile(FILE, code, serverOptions)).code).toMatchSnapshot();
    expect((await compile(FILE, code, clientOptions)).code).toMatchSnapshot();
  });
});
