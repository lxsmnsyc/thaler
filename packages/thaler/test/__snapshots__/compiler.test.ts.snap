// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`action$ > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$action as _$$action } from "thaler/server";
import { action$ } from 'thaler';
const _action$ = _$$action("/example/f0b3b6fa-index-0-example", async ({
  greeting,
  receiver
}) => {
  const message = greeting + ', ' + receiver + '!';
  return message;
});
const example = _$$clone(_action$);"
`;

exports[`action$ > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$action as _$$action } from "thaler/client";
import { action$ } from 'thaler';
const _action$ = _$$action("/example/f0b3b6fa-index-0-example");
const example = _$$clone(_action$);"
`;

exports[`custom server function > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$scope as _$$scope } from "thaler/server";
import { $$example as _$$example } from "example-server-function/server";
import { example$ } from 'example-server-function';
const _example$ = _$$example("/example/f0b3b6fa-index-0-example", () => {
  const [greeting, receiver] = _$$scope();
  const message = greeting + ', ' + receiver + '!';
  return message;
});
function exampleProgram() {
  const greeting = 'Hello';
  const receiver = 'World';
  const example = _$$clone(_example$, () => [greeting, receiver]);
}"
`;

exports[`custom server function > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$example as _$$example } from "example-server-function/client";
import { example$ } from 'example-server-function';
const _example$ = _$$example("/example/f0b3b6fa-index-0-example");
function exampleProgram() {
  const greeting = 'Hello';
  const receiver = 'World';
  const example = _$$clone(_example$, () => [greeting, receiver]);
}"
`;

exports[`fn$ > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$fn as _$$fn } from "thaler/server";
import { fn$ } from 'thaler';
const PREFIX = 'Message: ';
const _fn$ = _$$fn("/example/f0b3b6fa-index-0-example", ({
  greeting,
  receiver
}) => {
  const message = PREFIX + greeting + ', ' + receiver + '!';
  return message;
});
const example = _$$clone(_fn$, () => []);"
`;

exports[`fn$ > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$fn as _$$fn } from "thaler/client";
import { fn$ } from 'thaler';
const PREFIX = 'Message: ';
const _fn$ = _$$fn("/example/f0b3b6fa-index-0-example");
const example = _$$clone(_fn$, () => []);"
`;

exports[`fn$ > should transform with local scope 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$scope as _$$scope } from "thaler/server";
import { $$fn as _$$fn } from "thaler/server";
import { fn$ } from 'thaler';
const _fn$ = _$$fn("/example/f0b3b6fa-index-0-example", ({
  greeting,
  receiver
}) => {
  const [PREFIX] = _$$scope();
  const message = PREFIX + greeting + ', ' + receiver + '!';
  return message;
});
function test() {
  const PREFIX = 'Message: ';
  const example = _$$clone(_fn$, () => [PREFIX]);
}"
`;

exports[`fn$ > should transform with local scope 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$fn as _$$fn } from "thaler/client";
import { fn$ } from 'thaler';
const _fn$ = _$$fn("/example/f0b3b6fa-index-0-example");
function test() {
  const PREFIX = 'Message: ';
  const example = _$$clone(_fn$, () => [PREFIX]);
}"
`;

exports[`get$ > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$get as _$$get } from "thaler/server";
import { get$ } from 'thaler';
const _get$ = _$$get("/example/f0b3b6fa-index-0-example", ({
  greeting,
  receiver
}) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html'
    },
    status: 200
  });
});
const example = _$$clone(_get$);"
`;

exports[`get$ > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$get as _$$get } from "thaler/client";
import { get$ } from 'thaler';
const _get$ = _$$get("/example/f0b3b6fa-index-0-example");
const example = _$$clone(_get$);"
`;

exports[`loader$ > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$loader as _$$loader } from "thaler/server";
import { loader$ } from 'thaler';
const _loader$ = _$$loader("/example/f0b3b6fa-index-0-example", async ({
  greeting,
  receiver
}) => {
  const message = greeting + ', ' + receiver + '!';
  return message;
});
const example = _$$clone(_loader$);"
`;

exports[`loader$ > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$loader as _$$loader } from "thaler/client";
import { loader$ } from 'thaler';
const _loader$ = _$$loader("/example/f0b3b6fa-index-0-example");
const example = _$$clone(_loader$);"
`;

exports[`post$ > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$post as _$$post } from "thaler/server";
import { post$ } from 'thaler';
const _post$ = _$$post("/example/f0b3b6fa-index-0-example", ({
  greeting,
  receiver
}) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html'
    },
    status: 200
  });
});
const example = _$$clone(_post$);"
`;

exports[`post$ > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$post as _$$post } from "thaler/client";
import { post$ } from 'thaler';
const _post$ = _$$post("/example/f0b3b6fa-index-0-example");
const example = _$$clone(_post$);"
`;

exports[`pure$ > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$pure as _$$pure } from "thaler/server";
import { pure$ } from 'thaler';
const sleep = ms => new Promise(res => {
  setTimeout(res, ms, true);
});
const _pure$ = _$$pure("/example/f0b3b6fa-index-0-example", async ({
  greeting,
  receiver
}) => {
  await sleep(1000);
  const message = greeting + ', ' + receiver + '!';
  return message;
});
const example = _$$clone(_pure$);"
`;

exports[`pure$ > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$pure as _$$pure } from "thaler/client";
import { pure$ } from 'thaler';
const sleep = ms => new Promise(res => {
  setTimeout(res, ms, true);
});
const _pure$ = _$$pure("/example/f0b3b6fa-index-0-example");
const example = _$$clone(_pure$);"
`;

exports[`ref$ > should transform 1`] = `
"import { $$ref as _$$ref } from "thaler/server";
import { ref$ } from 'thaler';
const example = _$$ref("/example/f0b3b6fa-index-0", () => 'Hello World');"
`;

exports[`ref$ > should transform 2`] = `
"import { $$ref as _$$ref } from "thaler/client";
import { ref$ } from 'thaler';
const example = _$$ref("/example/f0b3b6fa-index-0", () => 'Hello World');"
`;

exports[`server$ > should transform 1`] = `
"import { $$clone as _$$clone } from "thaler/server";
import { $$server as _$$server } from "thaler/server";
import { server$ } from 'thaler';
const _server$ = _$$server("/example/f0b3b6fa-index-0-example", request => {
  return new Response('Hello World', {
    headers: {
      'content-type': 'text/html'
    },
    status: 200
  });
});
const example = _$$clone(_server$);"
`;

exports[`server$ > should transform 2`] = `
"import { $$clone as _$$clone } from "thaler/client";
import { $$server as _$$server } from "thaler/client";
import { server$ } from 'thaler';
const _server$ = _$$server("/example/f0b3b6fa-index-0-example");
const example = _$$clone(_server$);"
`;
